clients/vulnerable-code/src/main/kotlin/VulnerableCodeService.kt

```diff
@@ -0,0 +1,142 @@
+import com.fasterxml.jackson.databind.json.JsonMapper
+import com.fasterxml.jackson.module.kotlin.registerKotlinModule
+
+import okhttp3.OkHttpClient
+
+import retrofit2.Retrofit
+import retrofit2.converter.jackson.JacksonConverterFactory
+import retrofit2.http.Body
+import retrofit2.http.GET
+import retrofit2.http.POST
+import retrofit2.http.Path
+
+/**
+ * Interface for a REST service that allows interaction with the VulnerableCode API to query information about
+ * vulnerabilities detected for specific packages.
+ */
+interface VulnerableCodeService {
+    companion object {
+        /**
+         * Create a new service instance that connects to the [serverUrl] specified and uses the optionally provided
+         * [client].
+         */
+        fun create(serverUrl: String, client: OkHttpClient? = null): VulnerableCodeService {
+            val retrofit = Retrofit.Builder()
+                .apply { if (client != null) client(client) }
+                .baseUrl(serverUrl)
+                .addConverterFactory(JacksonConverterFactory.create(JsonMapper().registerKotlinModule()
+                    .setPropertyNamingStrategy(PropertyNamingStrategies.SNAKE_CASE)))
+                .build()
+
+            return retrofit.create(VulnerableCodeService::class.java)
+        }
+    }
+
+    /**
+     * Data class to represent a reference to a vulnerability. When querying the vulnerabilities for packages, each
+     * package in the result has a list of this type. In order to retrieve further details about vulnerabilities,
+     * an additional request has to performed.
+     */
+    data class HyperLinkedVulnerability(
+        /** A URL with information about the vulnerability. */
+        val url: String,
+
+        /** The ID of this vulnerability, typically a CVE identifier. */
+        @JsonProperty("vulnerability_id")
+        val vulnerabilityId: String?
+    )
+
+    /**
+     * Data class representing a vulnerability. Objects of this type are in the result of a query for vulnerabilities.
+     */
+    @JsonIgnoreProperties(ignoreUnknown = true)
+    data class Vulnerability(
+        /** A URL with information about the vulnerability. */
+        val url: String,
+
+        /**
+         * The identifier for this vulnerability. According to the VulnerableCode schema, this can be *null*.
+         */
+        val cveId: String?,
+
+        /**
+         * The vulnerability score. According to the VulnerableCode schema, this can be *null*.
+         */
+        val cvss: Float?
+    )
+
+    /**
+     * Data class describing a package in the result of a package query. An instance contains information about
+     * the vulnerabilities that have been found for this package.
+     */
+    @JsonIgnoreProperties(ignoreUnknown = true)
+    data class Vulnerabilities(
+        /** A list with vulnerabilities that have not yet been resolved. */
+        val unresolvedVulnerabilities: List<HyperLinkedVulnerability> = emptyList(),
```

Comment:
Why do we have a default value here and below, but in none of the other data classes?

Reply:
The lists are optional in the responses of VulnerableCode. Without the default values, parsing would fail if one of these lists would not be present.

Reply:
Should we then say "An optional list with ..." in both docs?